File: Desarrollo_de_apps/Practica1/C#/DoubleLinkedList/DoubleLinkedList/DoubleList.cs
using System;
namespace DoubleLinkedList
{
    public class DoubleList
    {
        //Atributos de la clase
        private Node head;
        private Node tail;

        //Constructor
        public DoubleList()
        {
            head = null;
            tail = null;
        }

        //Método para verificar si la lista está vacía
        public Boolean IsEmpty()
        {
            return ((head == null) && (tail == null));
        }

        //Mètodos para insertar un nodo al final
        public void InsertLast(string newValue)
        {
            InsertLast(new Node(newValue));
        }

        private void InsertLast(Node newNode)
        {
            if (IsEmpty())
            {
                head = newNode;
                tail = newNode;
            }
            else
            {
                tail.next = newNode;
                newNode.before = tail;
                tail = newNode;
            }
        }

        //Métodos para insertar un nodo al principio
        public void InsertBefore(string newValue)
        {
            InsertBefore(new Node(newValue));
        }

        private void InsertBefore(Node newNode)
        {
            if (IsEmpty())
            {
                head = newNode;
                tail = newNode;
            }
            else
            {
                newNode.next = head;
                head.before = newNode;
                head = newNode;
            }
        }

        //Métodos para insertar los nodos de manera ordenada según su valor
        public void InsertOrdered(string newValue)
        {
            InsertOrdered(new Node(newValue));
        }

        private void InsertOrdered(Node newNode)
        {
            if (IsEmpty())
            {
                head = newNode;
                tail = newNode;
            }
            else
            {
                //De acuerdo a cada compraración se decide en que lugar insertar el nodo
                if(newNode.CompareTo(head) < 0)
                {
                    InsertBefore(newNode);
                }
                else if(newNode.CompareTo(tail) > 0)
                {
                    InsertLast(newNode);
                }
                else
                {
                    Node behind = head;
                    Node after;

                    while (newNode.CompareTo(behind.next) > 0)
                    {
                        behind = behind.next;
                    }
                    after = behind.next;

                    newNode.next = behind.next;
                    newNode.before = behind;
                    behind.next = newNode;
                    after.before = newNode;
                }
            }
        }

        //Método para buscar un nodo con cierto valor en la lista
        public Node SerchNode(string value)
        {
            Node myNode = null;
            for(Node current = head; current != null; current = current.next)
            {
                if(string.Compare(value, current.value, StringComparison.Ordinal) == 0)
                {
                    myNode = current;
                    break;
                }
            }
            return myNode;
        }

        //Método para borrar el primer elemento de la lista
        private void DeleteFirst()
        {
            if(head != tail)
            {
                head = head.next;
                head.before = null;
            }
            else
            {
                head = null;
                tail = null;
            }
        }

        private void DeleteLast()
        {
            if(head != tail)
            {
                tail = tail.before;
                tail.next = null;
            }
            else
            {
                head = null;
                tail = null;
            }
        }
        
        private void DeleteAfter(Node nodeToDelete)
        {
            if(head != tail)
            {
                Node behind = nodeToDelete.before;
                Node after = nodeToDelete.next;

                behind.next = nodeToDelete.next;
                after.before = nodeToDelete.before;
            }
            else
            {
                head = null;
                tail = null;
            }
        }

        //Método par borrar un nodo de acuerdo a su valor
        public void Delete(string value)
        {
            if (!IsEmpty())
            {
                if(string.Compare(value, head.value, StringComparison.Ordinal) == 0)
                {
                    DeleteFirst();
                }
                else if(string.Compare(value, tail.value, StringComparison.Ordinal) == 0)
                {
                    DeleteLast();
                }
                else
                {
                    Node nodeToDelete = SerchNode(value);

                    if(nodeToDelete != null)
                    {
                        DeleteAfter(nodeToDelete);
                    }
                    else
                    {
                        Console.WriteLine("El valor no está en la lista!!!...");
                    }
                }
            }
            else
            {
                Console.WriteLine("La lista está vacía!!!...");
            }
        }

        //Método para imprimir la lista
        public void Print()
        {
            if (!IsEmpty())
            {
                for(Node current = head; current != null; current = current.next)
                {
                    Console.WriteLine(current.value);
                }
            }
            else
            {
                Console.WriteLine("La lista está vacía!!!...");
            }
        }

        //Método para imprimir en sentido contrario
        public void PrintReverse()
        {
            if (!IsEmpty())
            {
                for(Node current = tail; current != null; current = current.before)
                {
                    Console.WriteLine(current.value);
                }
            }
            else
            {
                Console.WriteLine("La lista está vacía!!!...");
            }
        }

        public void PrintVerbose()
        {
            if (!IsEmpty())
            {
                for(Node current = head; current != null; current = current.next)
                {
                    Console.WriteLine("Este: " + current.GetHashCode().ToString()
                     + "\tValor: " + current.value + "\tAnterior: " + (current.before 
                        != null ? current.before.GetHashCode().ToString() : "null 000")
                      + "\tSiguiente: " + (current.next != null ? current.next.GetHashCode().ToString()
                         : "null 000"));
                }
            }
            else
            {
                Console.WriteLine("La lista está vacía!!!...");
            }
        }

        public void PrintVerboseReverse()
        {
            if (!IsEmpty())
            {
                for(Node current = tail; current != null; current = current.before)
                {
                    Console.WriteLine("Este: " + current.GetHashCode().ToString()
                     + "\tValor: " + current.value + "\tAnterior: " + (current.before
                         != null ? current.before.GetHashCode().ToString() : "null 000")
                      + "\tSiguiente: " + (current.next != null ? current.next.GetHashCode().ToString()
                          : "null 000"));
                }
            }
            else
            {
                Console.WriteLine("La lista está vacía!!!...");
            }
        }
    }
}

File: Desarrollo_de_apps/Practica1/JAVA/DoubleLinkedList/src/doublelinkedlist/DoubleList.java

package doublelinkedlist;

public class DoubleList {
    //Atributos de la clase
    private Node head;
    private Node tail;
    
    //Constructor
    public DoubleList(){
        head = null;
        tail = null;
    }
    
    //Mètodo para validar si la lista està vacìa
    public boolean isEmpty(){
        return ((head == null) && (tail == null));
    }
    
    //Mètodos para insertar un nodo al final
    public void insertLast(String newValue){
        insertLast(new Node(newValue));
    }
    
    private void insertLast(Node newNode){
        if(isEmpty()){
            head = newNode;
            tail = newNode;
        }
        else{
            tail.next = newNode;
            newNode.before = tail;
            tail = newNode;
        }
    }
    
    //Mètodos para insertar un nodo al principio
    public void insertBefore(String newValue){
        insertBefore(new Node(newValue));
    }
    
    private void insertBefore(Node newNode){
        if(isEmpty()){
            head = newNode;
            tail = newNode;
        }
        else{
            newNode.next = head;
            head.before = newNode;
            head = newNode;
        }
    }
    
    //Mètodos para insertar un nodo en orden según su valor
    public void insertOrdered(String newValue){
        insertOrdered(new Node(newValue));
    }
    
    private void insertOrdered(Node newNode){
        if(isEmpty()){
            head = newNode;
            tail = newNode;
        }
        else{
            if(newNode.compareTo(head) < 0){
                insertBefore(newNode);
            }
            else if(newNode.compareTo(tail) > 0){
                insertLast(newNode);
            }
            else{
                Node behind = head;
                Node after;
                
                while(newNode.compareTo(behind.next) > 0){
                    behind = behind.next;
                }
                after = behind.next;
                
                newNode.next = behind.next;
                newNode.before = behind;
                behind.next = newNode;
                after.before = newNode;
            }
        }
    }
    
    //Mètodos para buscar un nodo por su valor
    public Node serchNode(String value){
        Node myNode = null;
        for(Node current = head; current != null; current = current.next){
            if(value.compareTo(current.value) == 0){
                myNode = current;
                break;
            }
        }
        return myNode;
    }
    
    //Método para borrar el primer nodo
    private void deleteFirst(){
        if(head != tail){
            head = head.next;
            head.before = null;
        }
        else{
            head = null;
            tail = null;
        }
    }
    
    //Método para borrar el último nodo
    private void deleteLast(){
        if(head != tail){
            tail = tail.before;
            tail.next = null;
        }
        else{
            head = null;
            tail = null;
        }
    }
    
    //Método para borrar nodo con cierto valor
    private void deleteAfter(Node nodeToDelete){
        if(head != tail){
            Node behind = nodeToDelete.before;
            Node after = nodeToDelete.next;
            
            behind.next = nodeToDelete.next;
            after.before = nodeToDelete.before;
        }
        else{
            head = null;
            tail = null;
        }
    }
    
    //Método para borrar un nodo en base a la comparación de su valor
    public void delete(String value){
        if(!isEmpty()){
            if(value.compareTo(head.value) == 0){
                deleteFirst();
            }
            else if(value.compareTo(tail.value) == 0){
                deleteLast();
            }
            else{
                Node nodeToDelete = serchNode(value);
                
                if(nodeToDelete != null){
                    deleteAfter(nodeToDelete);
                }
                else{
                    System.out.println("El valor no está en la lista!!!...");
                }
            }
        }
        else{
            System.out.println("La lista está vacía!!!...");
        }
    }
    
    //Método para imprimir la lista
    public void print(){
        if(!isEmpty()){
            for(Node current = head; current != null; current = current.next){
                System.out.println(current.value);
            }
        }
        else{
            System.out.println("La lista está vacía!!!...");
        }
    }
    
    //Método para imprimir la lista en orden inverso
    public void printReverse(){
        if(!isEmpty()){
            for(Node current = tail; current != null; current = current.before){
                System.out.println(current.value);
            }
        }
        else{
            System.out.println("La lista está vacía!!!...");
        }
    }
    
    public void printVerbose(){
        if(!isEmpty()){
            for(Node current = head; current != null; current = current.next){
                System.out.println("Este: " + Integer.toHexString(current.hashCode())
                        + "\tValor: " + current.value + "\tAnterior: " + (current.before
                                != null ? Integer.toHexString(current.before.hashCode()):
                        "null 000") + "\tSiguiente: " + (current.next
                                != null ? Integer.toHexString(current.next.hashCode()):
                                "null 000"));
            }
        }
        else{
            System.out.println("La lista está vacía!!!...");
        }
    }
    
    public void printVerboseReverse(){
        if(!isEmpty()){
            for(Node current = tail; current != null; current = current.before){
                System.out.println("Este: " + Integer.toHexString(current.hashCode())
                        + "\tValor: " + current.value + "\tAnterior: " + (current.before
                                != null ? Integer.toHexString(current.before.hashCode()):
                        "null 000") + "\tSiguiente: " + (current.next
                                != null ? Integer.toHexString(current.next.hashCode()):
                                "null 000"));
            }
        }
        else{
            System.out.println("La lista está vacía!!!...");
        }
    }
}












